src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionValueChoiceType.php

```diff
@@ -13,14 +13,24 @@
 
 namespace Sylius\Bundle\ProductBundle\Form\Type;
 
+use Sylius\Component\Product\Model\ProductInterface;
 use Sylius\Component\Product\Model\ProductOptionInterface;
+use Sylius\Component\Product\Resolver\AvailableProductOptionValuesResolverInterface;
 use Symfony\Component\Form\AbstractType;
 use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
 use Symfony\Component\OptionsResolver\Options;
 use Symfony\Component\OptionsResolver\OptionsResolver;
 
 final class ProductOptionValueChoiceType extends AbstractType
 {
+    /** @var AvailableProductOptionValuesResolverInterface */
+    private $availableProductOptionValuesResolver;
+
+    public function __construct(AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver)
```

Comment:
Why is this a BC break?

Anyway maybe we can keep it but with a setter injection. In the end it is used only when the `only_available_values` option is set and we can refactor to fallback on the old behavior if the `$availableProductOptionValuesResolver` property is not set.

Reply:
It is a BC Break, because one could redeclare class form type definition (for whatever reason) and with this change, the container can't be built. We just need to allow for null here and check if the argument is set before the execution of new logic.

Reply:
Ok so if I refactor with a setter injection, allowing the `$availableProductOptionValuesResolver` to be null, will you merge it sooner?

Reply:
I cannot merge it with the current solution for sure. Adding a setter would be a precedent and I'm not sure if there is any major benefit of it (compared to just allowing for nulling in the constructor). Service definition will be a little bit more complicated and we still need to ensure, that this value is set at the execution time. Nonetheless, I'm not against it, I'm just challenging an idea. 

```suggestion
    public function __construct(?AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver)
    {
        if (null === $availableProductOptionValuesResolver) {
            @trigger_error('Not passing availableProductOptionValuesResolver thru constructor is deprecated in Sylius 1.8 and it will be removed in Sylius 2.0');
        }
        
        $this->availableProductOptionValuesResolver = $availableProductOptionValuesResolver;
    }
```

Reply:
For optional dependencies I think it's better a setter injection. Taking from symfony docs (https://symfony.com/doc/current/service_container/calls.html):

> Usually, you'll want to inject your dependencies via the constructor. But sometimes, especially if a dependency is optional, you may want to use "setter injection".

But if you don't want a precedent I'm happy with the optional constructor argument as well. Let me know, I can work on it.

Reply:
Your point is valid, however, I would prefer to not have this argument as optional, but we need to keep it optional due to BC policy. If we will use setter injection, the upgrade path between 1.0 and 2.0 would be harder, as we would like to make it required in 2.0. With optional setter injection upgrade path would be smoother, as the same configuration will work on both versions. Let's go with optional constructor argument.

Reply:
Done in https://github.com/Sylius/Sylius/pull/11241/commits/572fa6aa5dd8fcd660b6bf11b15188213ddfc842